sql/privilege.go

```diff
@@ -165,16 +178,37 @@ func (p *PrivilegeDescriptor) Revoke(user string, privList privilege.List) {
 }
 
 // Validate is called when writing a descriptor.
-func (p *PrivilegeDescriptor) Validate() error {
+// If 'isSystem' is true, perform validation for system
+// database and tables.
+func (p *PrivilegeDescriptor) Validate(isSystem bool) error {
 	if p == nil {
 		return fmt.Errorf("missing privilege descriptor")
 	}
 	userPriv, ok := p.findUser(security.RootUser)
 	if !ok {
-		return fmt.Errorf("%s user does not have privileges", security.RootUser)
-	}
-	if userPriv.Privileges&(1<<privilege.ALL) == 0 {
-		return fmt.Errorf("%s user does not have ALL privileges", security.RootUser)
+		return fmt.Errorf("user %s does not have privileges", security.RootUser)
+	}
+	if isSystem {
+		// System databases and tables must have read-only permissions
+		// only (SELECT and GRANT). These must remain set for the root user.
+		var allowedPrivileges uint32
+		allowedPrivileges = (1 << privilege.SELECT) | (1 << privilege.GRANT)
+		if userPriv.Privileges&allowedPrivileges != allowedPrivileges {
+			return fmt.Errorf("user %s must have %s privileges on system objects",
+				security.RootUser, privilege.ListFromBitField(allowedPrivileges))
+		}
+
+		// For all users (root included), no other privileges must be granted.
+		for _, u := range p.Users {
+			if remaining := u.Privileges &^ allowedPrivileges; remaining != 0 {
+				return fmt.Errorf("user %s must not have %s privileges on system objects",
+					u.User, privilege.ListFromBitField(remaining))
+			}
+		}
+	} else if userPriv.Privileges&(1<<privilege.ALL) == 0 {
```

Comment:
this random bit-shifting in different places is awkward (ok, not going to start complaining about the definition of `privilege.X` again). `privilege.ALL.Mask()`?